IdentityCore/src/cache/accessor/MSIDDefaultTokenCacheAccessor.m

```diff
@@ -876,4 +884,42 @@ - (BOOL)saveAccount:(MSIDAccount *)account
     return tokens;
 }
 
+- (BOOL)saveAppMetadataWithConfiguration:(MSIDConfiguration *)configuration
+                            response:(MSIDTokenResponse *)response
+                             context:(id<MSIDRequestContext>)context
+                               error:(NSError **)error
+{
+    MSIDAppMetadataCacheItem *metadata = [_factory appMetadataFromResponse:response configuration:configuration];
+    metadata.environment = [[configuration.authority cacheUrlWithContext:context] msidHostWithPortIfNecessary];
+    
+    if (metadata)
+    {
+        return [self saveAppMetadata:metadata context:context error:error];
+    }
+   
+    return YES;
+}
+
+- (BOOL)saveAppMetadata:(MSIDAppMetadataCacheItem *)metadata
+            context:(id<MSIDRequestContext>)context
+              error:(NSError **)error
+{
+    MSIDTelemetryCacheEvent *event = [MSIDTelemetry startCacheEventWithName:MSID_TELEMETRY_EVENT_TOKEN_CACHE_WRITE context:context];
+    
+    BOOL result = [_accountCredentialCache saveAppMetadata:metadata context:context error:error];
+    [MSIDTelemetry stopCacheEvent:event withItem:nil success:result context:context];
+    
+    return result;
+}
+
+- (MSIDAppMetadataCacheItem *)getAppAppMetadataForConfiguration:(MSIDConfiguration *)configuration
+                                                        context:(id<MSIDRequestContext>)context
+                                                          error:(NSError *__autoreleasing *)error
+{
+    MSIDAppMetadataCacheKey *metadataKey = [[MSIDAppMetadataCacheKey alloc] initWithClientId:configuration.clientId
+                                environment:[[configuration.authority cacheUrlWithContext:context] msidHostWithPortIfNecessary]     generalType:MSIDAppMetadataType];
```

Comment:
authority has the "environment" property, please use that

Reply:
I believe that environment is not the same as the preferred_cache for instance, if login.microsoftonline.com is passed, should we save it with login.microsoftonline.com or login.windows.net